File: Algorithms/Sorting-Algorithm/Merge_Sort/non_recursive_merge_sort.py
def mergeSort(a):  
      
    current_size = 1
      
    while current_size < len(a) - 1:  
          
        left = 0
        while left < len(a)-1:  
              
            mid = min((left + current_size - 1),(len(a)-1)) 
              
            right = ((2 * current_size + left - 1,  
                    len(a) - 1)[2 * current_size  
                        + left - 1 > len(a)-1])  
                              
            merge(a, left, mid, right)  
            left = left + current_size*2
              
        current_size = 2 * current_size  
  
def merge(a, l, m, r):  
    n1 = m - l + 1
    n2 = r - m  
    L = [0] * n1  
    R = [0] * n2  
    for i in range(0, n1):  
        L[i] = a[l + i]  
    for i in range(0, n2):  
        R[i] = a[m + i + 1]  
  
    i, j, k = 0, 0, l  
    while i < n1 and j < n2:  
        if L[i] > R[j]:  
            a[k] = R[j]  
            j += 1
        else:  
            a[k] = L[i]  
            i += 1
        k += 1
  
    while i < n1:  
        a[k] = L[i]  
        i += 1
        k += 1
  
    while j < n2:  
        a[k] = R[j]  
        j += 1
        k += 1
  
  

a = list(map(int, input("Enter the numbers: ").split()))
print("Given list is ")  
print(a)  
  
mergeSort(a)  
  
print("Sorted list is ")  
print(a)  

File: Algorithms/Sorting-Algorithm/Merge_Sort/non_recursive_merge_sort.cs
using System; 
  
class MS { 
    static void mergeSort(int[] arr, 
                           int l, int r) 
    { 
        if (l < r) 
        { 
            int m = l + (r - l) / 2;  
            mergeSort(arr, l, m); 
            mergeSort(arr, m+1, r); 
            merge(arr, l, m, r); 
       } 
    } 
  
    static void merge(int[] arr, int l, 
                           int m, int r) 
    { 
        int i, j, k; 
        int n1 = m - l + 1; 
        int n2 = r - m; 
      
        int []L = new int[n1]; 
        int []R = new int[n2]; 
      

        for (i = 0; i < n1; i++) 
            L[i] = arr[l + i]; 
        for (j = 0; j < n2; j++) 
            R[j] = arr[m + 1+ j]; 
      
        i = 0; 
        j = 0; 
        k = l; 
        while (i < n1 && j < n2) 
        { 
            if (L[i] <= R[j]) 
            { 
                arr[k] = L[i]; 
                i++; 
            } 
            else
            { 
                arr[k] = R[j]; 
                j++; 
            } 
            k++; 
        } 
      
        while (i < n1) 
        { 
            arr[k] = L[i]; 
            i++; 
            k++; 
        } 

        while (j < n2) 
        { 
            arr[k] = R[j]; 
            j++; 
            k++; 
        } 
    } 
      
    static void printArray(int []A, int size) 
    { 
        int i; 
        for (i=0; i < size; i++) 
            Console.Write(A[i]+" "); 
        Console.Write("\n"); 
    } 
      
    public static void Main() 
    { 
        int []arr = {12, 11, 13, 5, 6, 7}; 
        int arr_size = arr.Length; 
      
        Console.Write("Given array is \n"); 
        printArray(arr, arr_size); 
      
        mergeSort(arr, 0, arr_size - 1); 
      
        Console.Write("\nSorted array is \n"); 
        printArray(arr, arr_size); 
    } 
} 